crates/revm/src/handler/optimism.rs

```diff
@@ -140,6 +144,133 @@ pub fn reward_beneficiary<SPEC: Spec, DB: Database>(
     Ok(())
 }
 
+/// Main return handle, returns the output of the transaction.
+#[inline]
+pub fn main_return<SPEC: Spec, DB: Database>(
+    data: &mut EVMData<'_, DB>,
+    call_result: InstructionResult,
+    output: Output,
+    gas: &Gas,
+) -> Result<ResultAndState, EVMError<DB::Error>> {
+    // used gas with refund calculated.
+    let gas_refunded = gas.refunded() as u64;
+    let final_gas_used = gas.spend() - gas_refunded;
+
+    // reset journal and return present state.
+    let (state, logs) = data.journaled_state.finalize();
+
+    let result = match call_result.into() {
+        SuccessOrHalt::Success(reason) => ExecutionResult::Success {
+            reason,
+            gas_used: final_gas_used,
+            gas_refunded,
+            logs,
+            output,
+        },
+        SuccessOrHalt::Revert => ExecutionResult::Revert {
+            gas_used: final_gas_used,
+            output: match output {
+                Output::Call(return_value) => return_value,
+                Output::Create(return_value, _) => return_value,
+            },
+        },
+        SuccessOrHalt::Halt(reason) => {
+            // Post-regolith, if the transaction is a deposit transaction and it haults,
+            // we bubble up to the global return handler. The mint value will be persisted
+            // and the caller nonce will be incremented there.
+            let is_deposit = data.env.tx.optimism.source_hash.is_some();
+            let optimism_regolith = data.env.cfg.optimism && SPEC::enabled(REGOLITH);
+            if is_deposit && optimism_regolith {
+                return Err(EVMError::Transaction(
+                    InvalidTransaction::HaltedDepositPostRegolith,
+                ));
+            }
+
+            ExecutionResult::Halt {
+                reason,
+                gas_used: final_gas_used,
+            }
+        }
+        SuccessOrHalt::FatalExternalError => {
+            return Err(EVMError::Database(data.error.take().unwrap()));
+        }
+        SuccessOrHalt::InternalContinue => {
+            panic!("Internal return flags should remain internal {call_result:?}")
+        }
+    };
+
+    Ok(ResultAndState { result, state })
```

Comment:
Would recommend doing something like this:
```
let ret = mainnet::main_return(..)?;
if let SuccessOrHalp::Halt(reason) = ret {


}
ret
```

This would reuse mainnet handler and just append optimism needed logic

Reply:
https://github.com/anton-rs/op-revm/pull/34

Reply:
Yeah this is much cleaner, nice idea.